magenta-lib/src/test/scala/magenta/deployment_type/S3ObjectPrefixParametersTest.scala

```diff
@@ -62,15 +67,77 @@ class S3ObjectPrefixParametersTest extends AnyFlatSpec with Matchers {
   }
 
   "prefixStagePaths" should "take precedent over other prefixes" in {
-    ???
+    val pkg = S3ObjectPrefixParametersTest.pkg(
+      Map(
+        "prefixStage" -> JsBoolean(false),
+        "prefixStack" -> JsBoolean(false),
+        "prefixPackage" -> JsBoolean(true),
+        "prefixApp" -> JsBoolean(false),
+        "prefixStagePaths" -> JsObject(
+          Seq(
+            "CODE" -> JsString("atoms-CODE"),
+            "PROD" -> JsString("atoms")
+          )
+        )
+      )
+    )
+    val path = TestS3DeploymentType.getPrefix(pkg, target, reporter)
+    path shouldBe "atoms"
   }
 
   "The S3 upload path" should "customisable via prefixStagePaths" in {
-    ???
+    val pkg = S3ObjectPrefixParametersTest.pkg(
+      Map(
+        "prefixStagePaths" -> JsObject(
+          Seq(
+            "CODE" -> JsString("atoms-CODE"),
+            "PROD" -> JsString("atoms")
+          )
+        )
+      )
+    )
+
+    // PROD stage deploy target
+    val productionPath = TestS3DeploymentType.getPrefix(pkg, target, reporter)
+    productionPath shouldBe "atoms"
+
+    // CODE stage deploy target
+    val codePath = TestS3DeploymentType.getPrefix(pkg, targetStageCODE, reporter)
+    codePath shouldBe "atoms-CODE"
   }
 
   "Setting prefixStagePaths" should "error when deploying to an unknown stage" in {
-    ???
+    val prefixFromStagePaths = Map(
+      "CODE" -> "atoms-CODE",
+      "PROD" -> "atoms"
+    )
+
+    val pkg = S3ObjectPrefixParametersTest.pkg(
+      Map(
+        "prefixStagePaths" -> JsObject(
+          prefixFromStagePaths map {case (stage, path) => (stage, JsString(path))}
+        )
+      )
+    )
+
+    val thrown = the[magenta.FailException] thrownBy {
+      // We expect using an stage deploy target to throw.
+      TestS3DeploymentType.getPrefix(pkg, targetStageUnknown, reporter)
+    }
+
+    thrown.getMessage should equal(
+      s"""
+         |Unable to locate prefix for stage ${targetStageUnknown.parameters.stage.name}.
+         |
+         |prefixFromStagePaths is set to:
+         |
+         |$prefixFromStagePaths
+         |
+         |To resolve, either:
+         |  - Deploy to a known stage
+         |  - Update prefixFromStagePaths, adding a value for ${targetStageUnknown.parameters.stage.name}
+         |""".stripMargin
```

Comment:
When used as a test assertion, I'd recommend using the raw string over an interpolated one as it is more explicit/clear what the expected message is, thus easier to sense check the wording.

Reply:
makes sense 👍  I usually go for concrete values in tests, but slipped my mind this time